### PR TITLE
EVG-5747 upload log files to S3

### PR DESCRIPTION
**Pull request by john-m-liu**

### CODE REVIEW

File: agent/agent.go [hunk 1]
```diff
@@ -37,6 +38,13 @@ type Options struct {
 	HeartbeatInterval  time.Duration
 	AgentSleepInterval time.Duration
 	Cleanup            bool
+	S3Opts             S3Config
+}
+
+type S3Config struct {
```
Comment (tychoish) [inline agent/agent.go]: can't we just use the S3Options from pail?

Comment (tychoish) [reply agent/agent.go]: also related, perhaps, but why is this public?

File: agent/logging.go [hunk 2]
```diff
@@ -87,33 +96,99 @@ func (a *Agent) makeLoggerProducer(ctx context.Context, c *model.LoggerConfig, t
...
 			SplunkServerURL:   taskConfig.SplunkServer,
 			SplunkToken:       taskConfig.SplunkToken,
-			Filepath:          fmt.Sprintf("%s/task.log", path),
+			Filepath:          fmt.Sprintf("%s/%s", a.opts.WorkingDirectory, taskLogs),
 		})
 	}
 	return a.comm.GetLoggerProducer(ctx, td, &config)
 }
+
+type s3UploadOpts struct {
```
Comment (tychoish) [inline agent/logging.go]: this feels like there's overlap with the other s3 options struct.

are we exposing configuration that sends logs to more than one bucket? or allow projects to specify different buckets for different logs. Do we even want that?

File: agent/logging.go [hunk 3]
```diff
@@ -87,33 +96,99 @@ func (a *Agent) makeLoggerProducer(ctx context.Context, c *model.LoggerConfig, t
...
+
+	return a.uploadLogFiles(ctx, tc, bucket, opts.bucket)
+}
+
+func (a *Agent) uploadLogFiles(ctx context.Context, tc *taskContext, bucket pail.Bucket, name string) error {
+	agentLogFile, err := os.Open(fmt.Sprintf("%s/%s", a.opts.WorkingDirectory, agentLogs))
+	if err != nil {
+		if !os.IsNotExist(err) {
+			grip.Error(errors.Wrap(err, "error opening agent log file"))
+		}
```
Comment (tychoish) [inline agent/logging.go]: maybe we can refactor this pattern to use some helper to check existance before opening the file? 

also the Bucket Interface has upload/download helpers which will read local files without needing to do os.Open

File: units/provisioning_agent_deploy.go [hunk 4]
```diff
@@ -298,6 +298,9 @@ func (j *agentDeployJob) startAgentOnRemote(ctx context.Context, settings *everg
 		fmt.Sprintf("--log_prefix='%s'", filepath.Join(hostObj.Distro.WorkDir, "agent")),
 		fmt.Sprintf("--working_directory='%s'", hostObj.Distro.WorkDir),
 		fmt.Sprintf("--logkeeper_url='%s'", settings.LoggerConfig.LogkeeperURL),
+		fmt.Sprintf("--s3_key='%s'", settings.Providers.AWS.Id),
+		fmt.Sprintf("--s3_secret='%s'", settings.Providers.AWS.Secret),
+		fmt.Sprintf("--s3_bucket='%s'", settings.Providers.AWS.Bucket),
```
Comment (tychoish) [inline units/provisioning_agent_deploy.go]: set these in environment variables, as this shows up in the output of ps, which we're piping to (in some cases) s3 which could be publicly visible.

File: agent/agent.go (hunk 1 above)
Comment (john-m-liu) [reply agent/agent.go]: changed, I had intended to customize this more but it wasn't needed

File: agent/logging.go (hunk 2 above)
Comment (john-m-liu) [reply agent/logging.go]: deleted

File: units/provisioning_agent_deploy.go (hunk 4 above)
Comment (john-m-liu) [reply units/provisioning_agent_deploy.go]: done

File: agent/logging.go (hunk 3 above)
Comment (john-m-liu) [reply agent/logging.go]: refactored so it's less verbose

File: units/provisioning_agent_deploy.go [hunk 5]
```diff
@@ -331,6 +331,10 @@ func (j *agentDeployJob) startAgentOnRemote(ctx context.Context, settings *everg
 		}
 	}
 
+	env["S3_KEY"] = settings.Providers.AWS.Id
+	env["S3_SECRET"] = settings.Providers.AWS.Secret
+	env["S3_BUCKET"] = settings.Providers.AWS.Bucket
```
[the remark this answers is not in the document]
Comment (john-m-liu) [reply units/provisioning_agent_deploy.go]: Split up the config into keys for S3 and EC2

**Pull request merged**